### PR TITLE
Added test and comments for poll_syscall

### PR DESCRIPTION
**Pull request by qianxichen233**
## Description

This PR added several tests and comments for poll_syscall. The test being added is very simliar to select_syscall, as both syscalls serve a very similar purpose. 

Fixes # (issue)

Several minor issues in poll_syscall are fixed in this PR:
1. POLLPRI flag should be used for exception fds instead of POLLERR
2. revents of all fds are cleared at the beginning of the syscall, in order to follow the standard
3. negative fds should be ignored
4. invalid fd should not raise the error directly, instead, it should just ignore it and set the corresponding flag in revent 

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Test A - `ut_lind_net_poll_bad_input`
- Test B - `ut_lind_net_poll_timeout`
- Test C - `ut_lind_net_poll` (this test is marked as ignore since an [unfixed bug](https://github.com/Lind-Project/safeposix-rust/issues/267) will panick the test)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: src/safeposix/syscalls/net_calls.rs [hunk 1]
```diff
@@ -2896,12 +2896,73 @@ impl Cage {
         return 0;
     }
 
+    /// ## ------------------POLL SYSCALL------------------
+    /// ### Description
+    /// poll_syscall performs a similar task to poll_syscall: it waits for
```
Comment (davidge20) [inline src/safeposix/syscalls/net_calls.rs]: poll_syscall performs a similar task to poll_syscall?

File: src/safeposix/syscalls/net_calls.rs [hunk 2]
```diff
@@ -2896,12 +2896,73 @@ impl Cage {
...
+    ///
+    /// ### Panics
+    /// No panic is expected from this syscall
     pub fn poll_syscall(
         &self,
         fds: &mut [PollStruct],
         timeout: Option<interface::RustDuration>,
     ) -> i32 {
-        //timeout is supposed to be in milliseconds
+        // current implementation of poll_syscall is based on select_syscall
```
Comment (davidge20) [inline src/safeposix/syscalls/net_calls.rs:2949]: Wow, interesting stuff!

File: src/safeposix/syscalls/net_calls.rs [hunk 3]
```diff
@@ -2896,12 +2896,73 @@ impl Cage {
...
+    /// ## ------------------POLL SYSCALL------------------
+    /// ### Description
+    /// poll_syscall performs a similar task to poll_syscall: it waits for
+    /// one of a set of file descriptors to become ready to perform I/O.
+
+    /// ### Function Arguments
+    /// The `poll_syscall()` receives two arguments:
+    /// * `fds` - The set of file descriptors to be monitored is specified in
+    ///   the fds argument, which is an array of PollStruct structures
+    ///   containing three fields: fd, events and revents. The field fd contains
```
Comment (davidge20) [inline src/safeposix/syscalls/net_calls.rs]: It's worth mentioning events and revents are requested events and returned events, respectively

File: src/safeposix/syscalls/net_calls.rs [hunk 4]
```diff
@@ -2911,34 +2972,52 @@ impl Cage {
...
+
+        // we loop until either timeout
+        // or any of the file descriptor is ready
         loop {
+            // iterate through each file descriptor
             for structpoll in &mut *fds {
+                // get the file descriptor
                 let fd = structpoll.fd;
+
+                // according to standard, we should ignore all file descriptor
```
Comment (davidge20) [inline src/safeposix/syscalls/net_calls.rs:2992]: How can file descriptors be less than 0?

File: src/safeposix/syscalls/net_calls.rs [hunk 5]
```diff
@@ -2947,20 +3026,40 @@ impl Cage {
...
 
+            // we break if there is any file descriptor ready
+            // or timeout is reached
             if return_code != 0 || interface::readtimer(start_time) > end_time {
                 break;
             } else {
+                // otherwise, check for signal and loop again
                 if interface::sigcheck() {
                     return syscall_error(Errno::EINTR, "poll", "interrupted function call");
                 }
```
Comment (davidge20) [inline src/safeposix/syscalls/net_calls.rs:3069]: What is lind_yield() on the line below?

File: src/tests/networking_tests.rs [hunk 6]
```diff
@@ -670,19 +670,248 @@ pub mod net_tests {
...
+        // this test is used for testing poll with some error/edge cases
+
+        // acquiring a lock on TESTMUTEX prevents other tests from running concurrently,
+        // and also performs clean env setup
+        let _thelock = setup::lock_and_init();
+
+        let cage = interface::cagetable_getref(1);
+
+        // error case 1: invalid file descriptor
+        // constrcut a PollStruct with invalid fd (10)
```
Comment (davidge20) [inline src/tests/networking_tests.rs]: contruct?

File: src/tests/networking_tests.rs [hunk 7]
```diff
@@ -670,19 +670,248 @@ pub mod net_tests {
...
+            0
+        );
+        // revents should be 0
+        assert_eq!(polled[0].revents, 0);
+
+        // edge case: revents should always be cleared
+        // create a file
+        let filefd = cage.open_syscall("/netpolltest.txt", O_CREAT | O_EXCL | O_RDWR, S_IRWXA);
+        assert!(filefd > 0);
+        // create a pipe
```
Comment (davidge20) [inline src/tests/networking_tests.rs:718]: How is the pipe being used here?

File: src/tests/networking_tests.rs [hunk 8]
```diff
@@ -670,19 +670,248 @@ pub mod net_tests {
...
+        // this test is used for testing poll with timeout behaviors specifically
+
+        // acquiring a lock on TESTMUTEX prevents other tests from running concurrently,
         // and also performs clean env setup
         let _thelock = setup::lock_and_init();
 
         let cage = interface::cagetable_getref(1);
 
+        // subtest 1: poll when timeout could expire
+        // create a TCP AF_UNIX socket
```
Comment (davidge20) [inline src/tests/networking_tests.rs]: AF_INET*

**Review by davidge20 — COMMENTED** [review]
Great tests! Left some minor comments for more context and typos

File: src/safeposix/syscalls/net_calls.rs (hunk 4 above)
Comment (qianxichen233) [reply src/safeposix/syscalls/net_calls.rs:2992]: a normal file descriptor could not be less than 0 of course. But the fd field is an user input, so it could be filled with anything that user may pass. And actually passing negative fd here in poll_syscall has a very common use, that it allows user to temporary disable monitoring this fd by just setting fd to -1.

File: src/safeposix/syscalls/net_calls.rs (hunk 5 above)
Comment (qianxichen233) [reply src/safeposix/syscalls/net_calls.rs:3069]: yield is an action in OS, that means "give up this cpu time, and let other process/thread do their work", which behaves like a very short period of sleep. We yield on each loop to make the file descriptor monitoring more efficient and not occupy too much time.

File: src/tests/networking_tests.rs (hunk 7 above)
Comment (qianxichen233) [reply src/tests/networking_tests.rs:718]: pipe fd is the most easy to use file descriptor (I think) that has a ready/unready state. Other types of file descriptors are either too complicated to create (e.g. socket fds), or just do not have an unready state (file fd).

File: src/safeposix/syscalls/net_calls.rs [hunk 9]
```diff
@@ -2911,34 +2976,52 @@ impl Cage {
...
         loop {
+            // iterate through each file descriptor
             for structpoll in &mut *fds {
+                // get the file descriptor
                 let fd = structpoll.fd;
+
+                // according to standard, we should ignore all file descriptor
+                // that is smaller than 0
+                if fd < 0 {
+                    continue;
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/net_calls.rs:2995]: Small doubt, here we are continuing if the fd is negative. Can you explain why we do this?

File: src/safeposix/syscalls/net_calls.rs (hunk 5 above)
Comment (namanlalitnyu) [reply src/safeposix/syscalls/net_calls.rs:3069]: Would be great to add this line in the form of comment here for understanding. 
`We yield on each loop to make the file descriptor monitoring more efficient and not occupy too much time.`

**Review by namanlalitnyu — COMMENTED** [review]
Looks good to me, just two minor comments. 
Very thorough tests!

File: src/safeposix/syscalls/net_calls.rs (hunk 9 above)
Comment (qianxichen233) [reply src/safeposix/syscalls/net_calls.rs:2995]: this is the behavior specified in standard, so we are just following it. Allowing this behavior has one convenient use that user could simply set the fd to negative to temporarily disable monitoring this file descriptor.

**Review by namanlalitnyu — APPROVED** [review]

**Comment by davidge20** [comment]
LGTM!

**Review by davidge20 — APPROVED** [review]

**Review by yashaswi2000 — APPROVED** [review]
LGTM, really great job!

File: src/safeposix/syscalls/net_calls.rs [hunk 10]
```diff
@@ -2947,23 +3030,45 @@ impl Cage {
...
+            // we break if there is any file descriptor ready
+            // or timeout is reached
             if return_code != 0 || interface::readtimer(start_time) > end_time {
                 break;
             } else {
+                // otherwise, check for signal and loop again
                 if interface::sigcheck() {
                     return syscall_error(Errno::EINTR, "poll", "interrupted function call");
                 }
+                // We yield on each loop to make the file descriptor monitoring more efficient
```
Comment (rennergade) [inline src/safeposix/syscalls/net_calls.rs]: I think this is better phrased as "We yield to let other threads continue if we've found no ready descriptors."

**Review by rennergade — CHANGES_REQUESTED** [review]
Minor comment but this is good to go. Great job!

**Review by rennergade — APPROVED** [review]
